Change ProcessGroupNCCL default timeout to 10 min

wconstab opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #110947

Avoid changing default for other backends as CPU backend (GLOO) may need
longer timeouts.

Motivated by trying to save cluster time when encountering collective
hangs.  Generally collectives should time out within seconds and 30
minutes (or 10 minutes) should provide ample headroom for edge cases.

wconstab committed:
Change ProcessGroupNCCL default timeout to 10 min

Avoid changing default for other backends as CPU backend (GLOO) may need
longer timeouts.

Motivated by trying to save cluster time when encountering collective
hangs.  Generally collectives should time out within seconds and 30
minutes (or 10 minutes) should provide ample headroom for edge cases.

[ghstack-poisoned]

wconstab committed:
Update on "Change ProcessGroupNCCL default timeout to 10 min"


Avoid changing default for other backends as CPU backend (GLOO) may need
longer timeouts.

Motivated by trying to save cluster time when encountering collective
hangs.  Generally collectives should time out within seconds and 30
minutes (or 10 minutes) should provide ample headroom for edge cases.

[ghstack-poisoned]

xw285cornell reviewed (approved):
Nice!

wconstab committed:
Update on "Change ProcessGroupNCCL default timeout to 10 min"


Avoid changing default for other backends as CPU backend (GLOO) may need
longer timeouts.

Motivated by trying to save cluster time when encountering collective
hangs.  Generally collectives should time out within seconds and 30
minutes (or 10 minutes) should provide ample headroom for edge cases.

[ghstack-poisoned]

fduwjj commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -1610,7 +1610,7 @@ void ProcessGroupNCCL::workEnqueue(
 
 ProcessGroupNCCL::Options::Options(bool is_high_priority_stream)
-    : Backend::Options(NCCL_BACKEND_NAME),
+    : Backend::Options(NCCL_BACKEND_NAME, kProcessGroupNCCLDefaultTimeout),
```
Why do we need this Options here?

wconstab replied:
we seem to have a subclass of Options for every single entity in c10d.  idk if it's a good pattern.

fduwjj commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.hpp:
```diff
@@ -176,7 +179,8 @@ class TORCH_API ProcessGroupNCCL : public Backend {
     bool checkTimeout(
         c10::optional<std::chrono::milliseconds> timeout = c10::nullopt);
-
+    bool checkDesyncTimeout(
+        c10::optional<std::chrono::milliseconds> timeout = c10::nullopt);
```
Is this used somewhere or do you plan to add two-tier timeout here?

wconstab replied:
ohh good catch this is not intended to land

wconstab committed:
Update on "Change ProcessGroupNCCL default timeout to 10 min"


Avoid changing default for other backends as CPU backend (GLOO) may need
longer timeouts.

Motivated by trying to save cluster time when encountering collective
hangs.  Generally collectives should time out within seconds and 30
minutes (or 10 minutes) should provide ample headroom for edge cases.

[ghstack-poisoned]

wconstab committed:
Update on "Change ProcessGroupNCCL default timeout to 10 min"


Avoid changing default for other backends as CPU backend (GLOO) may need
longer timeouts.

Motivated by trying to save cluster time when encountering collective
hangs.  Generally collectives should time out within seconds and 30
minutes (or 10 minutes) should provide ample headroom for edge cases.

[ghstack-poisoned]

fduwjj reviewed (approved):
LGTM.

wconstab commented:
@pytorchbot merge
